gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/JobInfo.java

```diff
@@ -0,0 +1,305 @@
+        @Override
+        public JobInfo apply(Job pb) {
+          return JobInfo.fromPb(pb);
+        }
+      };
+  private static final long serialVersionUID = -7086529810736715842L;
+
+  public enum CreateDisposition {
+    CREATE_IF_NEEDED,
+    CREATE_NEVER
+  }
+
+  public enum WriteDisposition {
+    WRITE_TRUNCATE,
+    WRITE_APPEND,
+    WRITE_EMPTY
+  }
+
+  private final String etag;
+  private final String id;
+  private final JobId jobId;
+  private final String selfLink;
+  private final JobStatus status;
+  private final StatisticsT statistics;
+  private final Boolean dryRun;
+  private final String userEmail;
+
+  public static abstract class Builder<T extends JobInfo, StatisticsT extends JobStatistics,
+      B extends Builder<T, StatisticsT, B>> {
+
+    private String etag;
+    private String id;
+    private JobId jobId;
+    private String selfLink;
+    private JobStatus status;
+    private StatisticsT statistics;
+    private Boolean dryRun;
+    private String userEmail;
+
+    protected Builder() {}
+
+    protected Builder(JobInfo<StatisticsT> jobInfo) {
+      this.etag = jobInfo.etag;
+      this.id = jobInfo.id;
+      this.jobId = jobInfo.jobId;
+      this.selfLink = jobInfo.selfLink;
+      this.status = jobInfo.status;
+      this.statistics = jobInfo.statistics;
+      this.dryRun = jobInfo.dryRun;
+      this.userEmail = jobInfo.userEmail;
+
+    }
+
+    protected Builder(Job jobPb) {
+      this.etag = jobPb.getEtag();
+      this.id = jobPb.getId();
+      if (jobPb.getJobReference() != null) {
+        this.jobId = JobId.fromPb(jobPb.getJobReference());
+      }
+      this.selfLink = jobPb.getSelfLink();
+      if (jobPb.getStatus() != null) {
+        this.status = JobStatus.fromPb(jobPb.getStatus());
+      }
+      if (jobPb.getStatistics() != null) {
+        this.statistics = JobStatistics.fromPb(jobPb.getStatistics());
+      }
+      if (jobPb.getConfiguration() != null) {
+        this.dryRun = jobPb.getConfiguration().getDryRun();
+      }
+      this.userEmail = jobPb.getUserEmail();
+    }
+
+    @SuppressWarnings("unchecked")
+    protected B self() {
+      return (B) this;
+    }
+
+    B etag(String etag) {
+      this.etag = etag;
+      return self();
+    }
+
+    B id(String id) {
+      this.id = id;
+      return self();
+    }
+
+    /**
+     * Sets the job identity.
+     */
+    public B jobId(JobId jobId) {
+      this.jobId = jobId;
+      return self();
+    }
+
+    B selfLink(String selfLink) {
+      this.selfLink = selfLink;
+      return self();
+    }
+
+    B status(JobStatus status) {
+      this.status = status;
+      return self();
+    }
+
+    B statistics(StatisticsT statistics) {
+      this.statistics = statistics;
+      return self();
+    }
+
+    B userEmail(String userEmail) {
+      this.userEmail = userEmail;
+      return self();
+    }
+
+    /**
+     * Sets whether the job has to be dry run or not. If set, the job is not executed. A valid query
```

Comment:
I wonder if we should have it only specified in `QueryJobInfo.Builder` (as I am not sure what is the behaviour of an "undefined behaviour" otherwise...)